fixes #12624 - add plugin interface to add provision methods

timogoebel opened the pull request:
This adds the possibility to add a provision_method via foreman's plugin api and show these methods in the gui.

timogoebel committed:
fixes #12624 - add plugin interface to add provision methods

dLobatog commented:
@timogoebel @ShimShtein started a discussion some time ago and is now implementing a way to do this through what we call Host facets. See https://groups.google.com/forum/?fromgroups#!topic/foreman-dev/esDULZDlOK8 for more info. 

If I'm not wrong adding new provisioning methods should be done through either extending the Provisioning facet or registering a new Provisioning facet, probably @ShimShtein or @jlsherrill can talk about more about it

domcleal commented:
> If I'm not wrong adding new provisioning methods should be done through either extending the Provisioning facet or registering a new Provisioning facet

I'd suggest this is reviewed anyway, since that doesn't exist.  It can be refactored later if somebody wants to while maintaining the plugin API.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -550,6 +555,19 @@ def self.count_habtm(association)
     association.camelize.constantize.find(counter.keys.compact).map {|i| {:label=>i.to_label, :data =>counter[i.id]}}
   end
 
+  def self.register_provision_method(name, friendly_name)
```
Ideally this should be on Foreman::Plugin (app/services/foreman/plugin.rb, test/unit/plugin_test.rb) so it can be added during plugin registration, then inside the list methods you can do something like appending the contents of: `Foreman::Plugin.all.map(&:provisioning_methods)` etc.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -2,7 +2,12 @@ class Host::Managed < Host::Base
+
+  class_attribute :provision_methods
+  self.provision_methods = {
+    'build' => 'Network Based',
```
Use `N_('Network Based')` to extract the strings without translating them (you're translating them correctly later with `_()` so that's good).

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -275,5 +275,10 @@ def apipie_ignored_controllers(controllers = nil)
+
+    # register a provision method
+    def provision_method(name, friendly_name)
+      Host::Managed.register_provision_method(name, friendly_name)
```
Oh sorry, I overlooked this!

domcleal replied:
In that case, prefer storing the registered provisioning inside this plugin object to ensure it stays in memory (in development mode Host::Managed may get reloaded) and then in Host::Managed retrieve and append the values from the registered plugins.

domcleal commented on app/views/hosts/_operating_system.html.erb:
```diff
@@ -9,26 +9,18 @@
   <%= field(f, :provision_method, :label => _('Provisioning Method'), :required => true) do
-      radio_button_f(f, :provision_method, :value=>'build', :checked=> @host.pxe_build?, :text=> _("Network Based"), :disabled => !@host.new_record?)+
-      radio_button_f(f, :provision_method, :value=>'image', :checked=> @host.image_build?, :text=> _("Image Based"), :disabled => !@host.new_record?)
+    Host::Managed::list_provision_methods.collect do |provision_method|
```
To save the extra `provision_method_friendly_name` method call, this could return the hash and iterate with `.each do |method,friendly_name|` ?

domcleal commented on app/models/host/managed.rb:
```diff
@@ -192,7 +197,7 @@ class Jail < ::Safemode::Jail
     validates :ptable_id, :presence => {:message => N_("can't be blank unless a custom partition has been defined")},
                           :if => Proc.new { |host| host.managed and host.disk.empty? and not Foreman.in_rake? and host.pxe_build? and host.build? }
-    validates :provision_method, :inclusion => {:in => PROVISION_METHODS, :message => N_('is unknown')}, :if => Proc.new {|host| host.managed?}
+    validates :provision_method, :inclusion => {:in => proc { self.provision_methods }, :message => N_('is unknown')}, :if => Proc.new {|host| host.managed?}
```
This could do with a unit test to check validation of a host works with a newly registered provisioning method.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -2,7 +2,12 @@ class Host::Managed < Host::Base
   include HostParams
-  PROVISION_METHODS = %w[build image]
+
+  class_attribute :provision_methods
```
To go with my comment about accessing the registered methods from the list of plugins, you could then skip the class attr and in the getter simply return `{'build'...}.merge(registered_methods)`.

timogoebel commented:
@domcleal : Thanks for reviewing, will implement your suggestions right away... :-)

timogoebel commented on app/assets/javascripts/host_edit.js:
```diff
@@ -75,24 +75,41 @@ function computeResourceSelected(item){
     build_provision_method_selected();
-  } else {
+  } else if(image) {
     $('#manage_network_build').hide();
```
We could delete this, as the next else case does the same. But that would make it harder to read. Any suggestions what to do?

domcleal replied:
I think I'm OK with deleting it, if only to reduce the number of branches.

timogoebel commented:
@domcleal : I implemented your suggestions and added the unit test. Mind to take a look?

ShimShtein commented:
@dLobatog, Facets were intended as a way to extend the host model. This feature does not add anything to the model object, so I don't think it's a good candidate for becoming a facet.

domcleal commented on app/assets/javascripts/host_edit.js:
```diff
@@ -75,24 +75,41 @@ function computeResourceSelected(item){
+      radio.click();
+      function_name = capabilities[0] + "_provision_method_selected";
+      if (typeof window[function_name] == 'function') {
+        eval(function_name+'()');
```
We should probably not add this but instead have plugins register event handlers on the radio boxes, which will make it more flexible.  We've got lots of examples of this (look for $(..).on in the codebase), so you could probably do something like `$('.host_provision_method_image').on('change', function() { .. })`

domcleal commented on app/views/hosts/_operating_system.html.erb:
```diff
@@ -9,26 +9,18 @@
   <%= field(f, :provision_method, :label => _('Provisioning Method'), :required => true) do
-      radio_button_f(f, :provision_method, :value=>'build', :checked=> @host.pxe_build?, :text=> _("Network Based"), :disabled => !@host.new_record?)+
-      radio_button_f(f, :provision_method, :value=>'image', :checked=> @host.image_build?, :text=> _("Image Based"), :disabled => !@host.new_record?)
+    @host.provision_methods.collect do |provision_method, friendly_name|
```
Consider sorting this either here in the view or in the model, else it might move around depending on how plugins register.

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1907,7 +1907,32 @@ def to_managed!
 
-  test "#image_build? must be true when provision_method is image" do
+  test "#provision_methods must include build and image by default" do
+    host = FactoryGirl.create(:host, :managed)
```
prefer `.build` rather than `.create` as it'll be instantiated and not saved to the database, speeding up the test

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1907,7 +1907,32 @@ def to_managed!
-  test "#image_build? must be true when provision_method is image" do
+  test "#provision_methods must include build and image by default" do
+    host = FactoryGirl.create(:host, :managed)
+    assert host.provision_methods.key?('build')
```
`assert_includes 'build', host.provision_methods.keys` helps as it'll print out the contents of the array if it fails

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1907,7 +1907,32 @@ def to_managed!
+  end
+
+  test "#registered_provision_methods must return empty hash if no plugins are initialized" do
+    host = FactoryGirl.create(:host, :managed)
```
ditto, use .build

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1907,7 +1907,32 @@ def to_managed!
+
+  test "#registered_provision_methods must return empty hash if no plugins are initialized" do
+    host = FactoryGirl.create(:host, :managed)
+    assert_equal host.registered_provision_methods, {}
```
swap the arguments, `expected, actual`

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1907,7 +1907,32 @@ def to_managed!
+      name 'Awesome provision'
+      provision_method 'awesome', 'Awesomeness Based'
+    end
+    host = FactoryGirl.create(:host,
```
ditto, use .build and then you can also move the stub to `host.expects(:capabilities)..` (I think you're using any_instance because creating the object will validate it in the process)

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1907,7 +1907,32 @@ def to_managed!
+                              :with_environment,
+                              :provision_method => 'awesome'
+                             )
+    assert(host.valid?)
```
`assert_valid host`

domcleal commented on test/unit/plugin_test.rb:
```diff
@@ -260,4 +260,13 @@ def test_register_custom_status
+      name 'Awesome provision'
+      provision_method 'awesome', 'Awesomeness Based'
+    end
+    host = FactoryGirl.create(:host, :managed)
```
build

domcleal commented:
1) Please take a look at the tests, I think there's a file with some bad syntax (wrong end block probably).

2) The API controller (app/controllers/api/v2/hosts_controller.rb) lists provision_method but doesn't say which value(s) it accepts.  I'd suggest adding a description and then use the new Host::Managed methods to fill it in.

Look at how :providers is set in config/initializers/apipie.rb and used in app/controllers/api/v2/compute_resources_controller.rb's :provider parameter and copy that for provision_method.

3) The Host::Managed#capabilities method might be limiting too, as it defers to the compute resource for CR-based hosts and is hard coded to [:build] in bare metal cases.

I think for your bootdisk change you'd probably need to extend the CR models, but I'm wondering if we need a more standard/pluggable way to add the method to both CR and bare metal capabilities.  Any thoughts?

timogoebel commented:
@domcleal:

> 1) Please take a look at the tests, I think there's a file with some bad syntax (wrong end block probably).

I lost a 'test' somewhere, don't know how that happened... Thanks for the hints in the inline comments. Improved the tests.

> 2) The API controller (app/controllers/api/v2/hosts_controller.rb) lists provision_method but doesn't say which value(s) it accepts. I'd suggest adding a description and then use the new Host::Managed methods to fill it in.

Done. Thanks for the hint, did not not that possibility existed.

> 3) The Host::Managed#capabilities method might be limiting too, as it defers to the compute resource for CR-based hosts and is hard coded to [:build] in bare metal cases.

I did not think of that. Of course for the boot disk iso change, I also want to extend this. I refactored Host::Managed.capabilities to call Host::Managed.bare_metal_capabilities. This can easily be extended by an alias_method_chain. That's probably not the best solution, but this change is already bigger than I imagined at first. For the UI part, I added a hidden field that stores the value of bare_metal_capabilities and query that via jQuery instead of using the hard coded ':build'. What do you think?

domcleal commented on config/initializers/apipie.rb:
```diff
@@ -20,6 +20,7 @@
     :providers => ComputeResource.providers.join(', '),
     :default_nic_type => InterfaceTypeMapper::DEFAULT_TYPE.humanized_name.downcase,
     :template_kinds => -> { TemplateKind.scoped.map(&:name).join(", ") },
+    :provision_methods => Host::Managed.provision_methods.map { |method, friendly_name| "#{method} (#{friendly_name})" }.join(', '),
```
friendly_name here won't get translated into the per-language API docs.  You could leave it out or change this to a lambda so it's run on the fly, so it can be translated correctly: `-> { Host::Managed.provision_methods.map { |method, friendly_name| "#{method} (#{_(friendly_name)})" }.join(', ') },`

timogoebel replied:
Let's keep this simple, will change it to just display the keys.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -550,6 +549,17 @@ def self.count_habtm(association)
   end
 
+  def self.provision_methods
+    Hash[{
```
I don't think converting the sorted array into a hash again will preserve the sort order, might be safer to do it in the UI (or wherever it's used).

domcleal commented:
> > 3) The Host::Managed#capabilities method might be limiting too, as it defers to the compute resource for CR-based hosts and is hard coded to [:build] in bare metal cases.
> 
> I did not think of that. Of course for the boot disk iso change, I also want to extend this. I refactored Host::Managed.capabilities to call Host::Managed.bare_metal_capabilities. This can easily be extended by an alias_method_chain. That's probably not the best solution, but this change is already bigger than I imagined at first. For the UI part, I added a hidden field that stores the value of bare_metal_capabilities and query that via jQuery instead of using the hard coded ':build'. What do you think?

Yes, I think that's OK.  My only other thoughts were to add it as a data attribute on the provision_method input field, but that's not always present, or to add it to the form tag itself, but I think it's fine as-is.

timogoebel commented:
@domcleal: Refactored the sorting to the ui and implemented your lambda. Mind to take another look?
I think, the code quality improved a lot because of your annotations. Just wanted to say thanks.

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -275,5 +276,11 @@ def apipie_ignored_controllers(controllers = nil)
+    # register a provision method
+    def provision_method(name, friendly_name)
+      return if @provision_methods.key?(name)
+      @provision_methods[name] = friendly_name
```
It might be best to convert `name` to `name.to_s` on entry to this method, I instinctively wrote `provision_method :test, ..` and it threw a runtime error later when viewing the New Host form.

domcleal commented on app/views/hosts/_operating_system.html.erb:
```diff
@@ -9,26 +9,18 @@
 
 <div id='provisioning_method' <%= display? (@host.capabilities.size < 2)  %> >
   <%= field(f, :provision_method, :label => _('Provisioning Method'), :required => true) do
```
add `:size => 'col-md-10'` in here to make the field full width, this will stop ugly wrapping of the radio buttons when adding new methods:

![screenshot from 2015-12-07 14-25-26](https://cloud.githubusercontent.com/assets/482089/11629157/69c85574-9cee-11e5-89a5-746640ac4442.png)

domcleal commented:
> I think, the code quality improved a lot because of your annotations. Just wanted to say thanks.

No problem, that's the aim!

timogoebel commented:
@domcleal : Great. Updated the size and converted "name" to String.

domcleal commented:
Merged as 51685ba, thanks @timogoebel.

If you get the opportunity, please could you add a description to http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin (with a >= 1.11 note)?
